### ISSUE TITLE
Number and Nested attributes not implemented

### ISSUE THREAD

**Issue by robertjustjones** [aveine/JsonAPI#6]
I've realized that nested attributes in the JsonAPI resources aren't implemented. I've implemented RequestTests to demonstrate what I would expect to work, including the aliasing of attribute names. I'll do a PR for just the tests.

Have you given thought to implementing this already?

**Comment by robertjustjones** [comment]
I now see that numeric values are not implemented either. I've updated the tests in the PR to include those, both top-level and nested.

**Comment by Tazrof** [comment]
Hello !

From my point of view numbers and nested attributes are implemented and tested, see:
- Test class definition: https://github.com/aveine/JsonAPI/blob/b64b385c17cb2a81837ee261554fd760905a1bfa/Tests/JsonAPITests/TestClasses.swift#L5 
- Serialization test: https://github.com/aveine/JsonAPI/blob/b64b385c17cb2a81837ee261554fd760905a1bfa/Tests/JsonAPITests/ResourceTests.swift#L136
- Unserialization test: https://github.com/aveine/JsonAPI/blob/b64b385c17cb2a81837ee261554fd760905a1bfa/Tests/JsonAPITests/ResourceTests.swift#L217
- https://github.com/aveine/JsonAPI/runs/3557162005?check_suite_focus=true latest CI is ok about the tests

We use this library in production with numbers and nested attributes in payloads and don't have any issues with it. Do you have some issues on your side? If you have a payload that doesn't work on your side I can check if something goes wrong with the library.

I did review your PR and I think the library could indeed improve nested attributes management with the possibility to override nested keys in `resourceAttributesKeys`.

**Comment by robertjustjones** [comment]
That's great then. I understand how the resourceAttributesKeys complicates nested. 

So excepting the "town"/"city" mapping, why don't the tests in #7  pass?  I'm seeing `nil` for all numbers and the nested fields.  Maybe I've configured something incorrectly?

**Comment by Tazrof** [comment]
So I've lookup in depth your PR this morning to see the problems.

The main problem is that since Swift doesn't have any internal reflection library that allow to update values (Swift `Mirror` is read only), I've bypassed the limitation by inheriting `NSObject` in `Resource` which allow to read/write values in a generic way with `NSKeyValueCoding`. However this come with a major limitation: only objective-c types can be manipulated. Therefore if you declare an attribute that is not an objective-c type (like an `Int`), it won't work. That's why the `ComplexResource` in https://github.com/aveine/JsonAPI/blob/b2b28e5aab8cd216dc8fd22551e07fc198a00531/Tests/JsonAPITests/TestClasses.swift use `NSNumber` and not `Int` or `Double`.

For the nested attributes part the problem is a little bit different but still linked. Internally we never had nested attributes that was "complex" so that we created a sub class/struct to define the properties (when it is the case, it is usually a relationship on another model). Instead we currently declare the nested attributes as a `NSDictionary` and populate some other attributes with the property observer `didSet` like this:
```swift
class MyResource: Resource {
    var descriptions: NSDictionary? {
        didSet {
            let lang = Locale.current.languageCode ?? "en"
            self.desc = descriptions?[lang] as? String ?? descriptions?["en"] as? String
        }
    }
    var desc: String?
 }
```
So there is definitely room for improvement here, like what you did in the PR with the struct definition (which is the cleanest way, and also more type safe).
But there is also another problem even if I implement the feature: structs are Swift types with no equivalent in objective-c. Therefore, as described above, trying to use this type in a resource definition won't work. Classes would work, like it is the case for relationship actually.

One way to solve the issue around objective-c limitation could be to drop the ease of using `NSObject` and rely on an external reflection library since I won't have the time to implement the mechanisms internally by playing with pointers. This way it should allow the use of Swift type.

My plan is the following:
- Try implementing https://github.com/wickwirew/Runtime to set the resource's attributes in the unserialization process instead of using the objective-c layer. It should be rather quick to implement from what I see and since the tests are quite exhaustive it should give a clear green or red light.
- Add the possibility to use struct/class to define nested attributes in a cleaner and type safe way.
- Add the possibility to override nested keys in `resourceAttributesKeys` as discussed before.

The minor inconvenience is that I don't have access to a macOS environment outside work anymore and right now there are a lot of things to do, so I don't have a clear timeline for the implementation. 
I will try to give you an update before the end of the week, I should manage to get some work done on it.

**Comment by robertjustjones** [comment]
Yes, it looked rather involved to me, as well, as I dug into it.

**Comment by robertjustjones** [comment]
@Tazrof you might check out the [Japx](https://github.com/infinum/Japx) implementation.

**Comment by Tazrof** [comment]
Hello ! Happy new year !

Sorry to came back to you after so long, it has been a crazy end of year at work.
We are currently cleaning some parts of our technical stack as well as updating our codebase for the beginning of the year. As so, I've implemented the first bullet point of my last message and now the library can use native Swift type for Bool, Int and Double! It is already available with the new 1.1.0 version.

I will see if I have more time for the last two points this month and allows the use of nested attributes more type safely, but I think I can manage something with Codable quite quickly.

**Comment by Tazrof** [comment]
Hello !

At long last I've implemented the nested attributes handling with version 1.3.0 . The struct / class just need to inherit from the protocol `ResourceNestedAttribute` ([documentation](https://aveine.github.io/JsonAPI/Protocols/ResourceNestedAttribute.html#/s:7JsonAPI23ResourceNestedAttributeP20nestedAttributesKeysSDyS2SGvpZ)), and voila !

There is an example in the README [here](https://aveine.github.io/JsonAPI/index.html#usage), or even in the classes for the tests [here](https://github.com/aveine/JsonAPI/blob/master/Tests/JsonAPITests/TestClasses.swift#L121).

I've moved the handling of the remapping and the exclusions of the attributes from the `Resource` class, as you've suggested in your PR, into the new protocol to tackle issues where you may have wanted to remap the attribute itself and some of its nested attributes. Moreover, I think it is more clear and may allow some factorisation in some cases.

If you have any questions and/or feedback, don't hesitate !
